core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp

```diff
@@ -0,0 +1,458 @@
+    for (auto i1 = begin1; i1 < end1; ++i1) {                                 \
+      for (auto i0 = begin0; i0 < end0; ++i0) {                               \
+        functor(i0, i1, val);                                                 \
+      }                                                                       \
+    }                                                                         \
+    aval = val;                                                               \
+  }                                                                           \
+                                                                              \
+  template <class ValueType, class Functor>                                   \
+  void OpenACCParallelReduce##REDUCER(OpenACCIterateRight, ValueType& aval,   \
+                                      Functor const& afunctor,                \
+                                      OpenACCMDRangeBegin<2> const& begin,    \
+                                      OpenACCMDRangeEnd<2> const& end,        \
+                                      int async_arg) {                        \
+    auto val = aval;                                                          \
+    auto const functor(afunctor);                                             \
+    int begin0 = begin[0];                                                    \
+    int end0   = end[0];                                                      \
+    int begin1 = begin[1];                                                    \
+    int end1   = end[1];                                                      \
+    /* clang-format off */ \
+    KOKKOS_IMPL_ACC_PGRAMA(parallel loop gang vector collapse(2) reduction(OPERATOR:val) copyin(functor) async(async_arg))                                                  \
+    /* clang-format on */                                                     \
+    for (auto i0 = begin0; i0 < end0; ++i0) {                                 \
+      for (auto i1 = begin1; i1 < end1; ++i1) {                               \
+        functor(i0, i1, val);                                                 \
+      }                                                                       \
+    }                                                                         \
+    aval = val;                                                               \
+  }                                                                           \
+                                                                              \
+  template <class ValueType, class Functor>                                   \
+  void OpenACCParallelReduce##REDUCER(OpenACCIterateLeft, ValueType& aval,    \
+                                      Functor const& afunctor,                \
+                                      OpenACCMDRangeBegin<3> const& begin,    \
+                                      OpenACCMDRangeEnd<3> const& end,        \
```

Comment:
Can we add the rank as a template parameter to the function call? 
`template <int rank, class ValueType, class Functor>`

Reply:
How would that help?  You still have to explicitly write unpack write the loops.

Reply:
Yes, we will have to write the loops explicitly but I was thinking of something like this 
```
template <int Rank, class ValueType, class Functor>
 void std::enable_if_t<Rank == 3> OpenACCParallelReduce(.....) {...}
```
Essentially something to reduce the number of parameters in the function call. But this is not going to help I guess.

Reply:
It does not reduce the numbers of parameters